Source/SwiftLintFramework/Rules/ExplictSelfRule.swift

```diff
@@ -0,0 +1,132 @@
+            let bodyLocation = (dictionary["key.bodyoffset"] as? Int64).flatMap({ Int($0) }),
+            let bodyLength = (dictionary["key.bodylength"] as? Int64).flatMap({ Int($0) })
+            else { return [] }
+
+        //get all the tokens inside the body
+        let range = NSRange(location: bodyLocation, length: bodyLength)
+        let tokens = file.syntaxMap.tokensIn(range)
+
+        var violations: [StyleViolation] = []
+
+        var previous: [String] = []
+        for token in tokens {
+            guard let type = SyntaxKind(rawValue: token.type) else { continue }
+
+            let value = file.contents.substring(token.offset, length: token.length)
+            defer { previous.append(value) }
+
+            let allowedSuffixes: [[String]] = [
+                ["self"], //self.value (normal access)
+                ["if", "let"] //if let value = self.value (shadowing)
+            ]
+
+            let isInstanceIdentifier = (type == .Identifier && instanceMemberNames.contains(value))
+
+            if isInstanceIdentifier && !previous.suffix(matches: allowedSuffixes) {
+                violations.append(StyleViolation(
+                    ruleDescription: self.dynamicType.description,
+                    severity: configuration.severity,
+                    location: Location(file: file, byteOffset: token.offset)
+                    )
+                )
+            }
+        }
+
+        return violations
+    }
+}
+
+extension File {
+    func members(declarations: [SwiftDeclarationKind]) -> [[String: SourceKitRepresentable]] {
+        return self.members(self, dictionary: self.structure.dictionary, declarations: declarations)
+    }
+    private func members(
+        file: File,
+        dictionary: [String: SourceKitRepresentable],
+        declarations: [SwiftDeclarationKind]) -> [[String: SourceKitRepresentable]] {
+
+        let substructure = dictionary["key.substructure"] as? [SourceKitRepresentable] ?? []
+
+        return substructure.flatMap { subItem -> [[String: SourceKitRepresentable]] in
+            guard let subDict = subItem as? [String: SourceKitRepresentable],
+                kindString = subDict["key.kind"] as? String,
+                kind = SwiftDeclarationKind(rawValue: kindString)
+                else { return [] }
+
+            return self.members(file, dictionary: subDict, declarations: declarations) +
+                (declarations.contains(kind) ? [subDict] : [])
+        }
+    }
+}
+
+extension SequenceType where Generator.Element: Equatable {
+    func suffix(matches items: [Generator.Element]) -> Bool {
+        let suffix = Array(self.suffix(items.count))
+
+        // swiftlint:disable control_statement
+        for (a, b) in zip(suffix, items) {
+            if a == b { return true }
+        }
+        return false
+    }
+
+    func suffix(matches items: [[Generator.Element]]) -> Bool {
+        for test in items
+            where self.suffix(matches: test) { return true }
```

Comment:
perhaps I'm misunderstanding what you mean? this function takes an `Array<Array<Generator.Element>>` and neither this function or the one it calls is equivalent to `.contains` - they are both checking the tail of the array for a set of elements in a specific order. can you perhaps show me a snippet of what you mean?